src/uct/tcp/tcp_ep.c

```diff
@@ -62,64 +74,211 @@ static void uct_tcp_ep_epoll_ctl(uct_tcp_ep_t *ep, int op)
 
 static inline int uct_tcp_ep_can_send(uct_tcp_ep_t *ep)
 {
-    ucs_assert(ep->offset <= ep->length);
+    ucs_assert(ep->tx.offset <= ep->tx.length);
     /* TODO optimize to allow partial sends/message coalescing */
-    return ep->length == 0;
+    return ep->tx.length == 0;
+}
+
+static ucs_status_t uct_tcp_ep_ctx_init(uct_tcp_iface_t *iface,
+                                        uct_tcp_ep_ctx_t *ctx)
+{
+    ctx->buf = ucs_malloc(ucs_max(iface->config.buf_size,
+                                  iface->config.short_size), "tcp_buf");
+    if (ctx->buf == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    ctx->offset = 0;
```

Comment:
we can add wrapper for ucx_tcp_ep_ctx_rewind() to set length and offset to 0. can be also called from uct_tcp_ep_send()